firedrake/solving_utils.py

```diff
@@ -21,6 +23,41 @@ def _make_reasons(reasons):
 SNESReasons = _make_reasons(PETSc.SNES.ConvergedReason())
 
 
+if get_config()["options"]["petsc_int_type"] == "int32":
+    DEFAULT_KSP_PARAMETERS = {"mat_type": "aij",
+                              "ksp_type": "preonly",
+                              "pc_type": "lu",
+                              "pc_factor_mat_solver_type": "mumps"}
+else:
+    # FIXME: which factorisation package to use in int64?
+    DEFAULT_KSP_PARAMETERS = {}
+
+
+def set_defaults(solver_parameters, arguments, *defaults):
+    """Set defaults for solver parameters.
+
+    :arg solver_parameters: dict of user solver parameters to override/extend defaults
+    :arg arguments: arguments for the bilinear form (need to know if we have a Real block).
+    :arg defaults: Any default sets of parameters to supply."""
+    parameters = dict(chain.from_iterable(d.items() for d in defaults))
+    if any(isinstance(V, fsi.RealFunctionSpace)
+           for V in chain.from_iterable(a.function_space() for a in arguments)):
+        # There's a Real block, override with defaults with nest + jacobi.
+        parameters.update({"pc_type": "jacobi", "mat_type": "nest"})
```

Comment:
I think in this case we should warn the user and generate a set of options that takes a full Schur fieldsplit onto the R-blocks and uses LU for the remainder (à la the Moore-Spence tests in firedrake). This is the most robust solver we can automatically pick, in the absence of assembled R blocks.